examples/next-prisma-sse-subscriptions/src/server/routers/post.ts

```diff
@@ -0,0 +1,194 @@
+        orderBy: {
+          createdAt: 'desc',
+        },
+        cursor: cursor ? { createdAt: cursor } : undefined,
+        take: take + 1,
+        skip: 0,
+      });
+      const items = page.reverse();
+      let nextCursor: typeof cursor | null = null;
+      if (items.length > take) {
+        const prev = items.shift();
+        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+        nextCursor = prev!.createdAt;
+      }
+      return {
+        items,
+        nextCursor,
+      };
+    }),
+
+  onAdd: publicProcedure
+    .input(
+      z.object({
+        lastEventId: z.string().nullish(),
+      }),
+    )
+    .subscription(async function* (opts) {
+      async function getPostsSince(date: Date | null) {
+        const where: Prisma.PostWhereInput = date
+          ? {
+              createdAt: {
+                gt: date,
+              },
+            }
+          : {};
+
+        return await prisma.post.findMany({
+          where,
+          orderBy: {
+            createdAt: 'asc',
+          },
+        });
+      }
+
+      let lastMessageCursor: Date | null = null;
+
+      if (opts.input.lastEventId) {
+        const itemById = await prisma.post.findUnique({
+          where: {
+            id: opts.input.lastEventId,
+          },
+        });
+        lastMessageCursor = itemById?.createdAt ?? null;
+      }
+
+      let unsubscribe = () => {
+        //
+      };
+      const stream = new ReadableStream<Post>({
+        async start(controller) {
+          const onAdd = (data: Post) => {
+            controller.enqueue(data);
+          };
+          dbEvents.on('add', onAdd);
+
+          const items = await getPostsSince(lastMessageCursor);
+          for (const item of items) {
+            controller.enqueue(item);
+          }
+          unsubscribe = () => {
+            dbEvents.off('add', onAdd);
+          };
+        },
+        cancel() {
+          unsubscribe();
+        },
+      });
+
+      for await (const post of streamToAsyncIterable(stream)) {
+        yield {
+          id: post.id,
+          data: post,
+        } satisfies SSEvent;
+      }
+    }),
+
+  whoIsTyping: publicProcedure
+    .input(
+      z
+        .object({
+          lastEventId: z.string().optional(),
+        })
+        .optional(),
+    )
+    .subscription(async function* (opts) {
+      let lastEventId = opts?.input?.lastEventId ?? '';
+
+      const maybeYield = function* (who: WhoIsTyping) {
+        const id = Object.keys(who).sort().toString();
+        if (lastEventId === id) {
+          return;
+        }
+        yield {
+          id,
+          data: Object.keys(who),
+        } satisfies SSEvent;
+
+        lastEventId = id;
+      };
+
+      // if someone is typing, emit event immediately
+      yield* maybeYield(currentlyTyping);
+
+      for await (const [who] of dbEvents.toIterable('isTypingUpdate')) {
+        yield* maybeYield(who);
+      }
```

Comment:
Are you using this just to test out iterables or does this pattern have benefits over using observable and eventListeners in this situation?